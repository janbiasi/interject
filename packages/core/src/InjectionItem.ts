import { LookupFlags } from './Flags';

export interface IInjection {
	token: any;
	factory: Function;
	useNew: boolean;
	dependencies: IInjectionDependency[];
	value: any;
	useExisting: any;
}

export interface IInjectionDependency {
	token: any;
	options: LookupFlags;
}
