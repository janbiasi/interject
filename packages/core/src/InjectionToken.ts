import { ProviderScope } from './Scope';
import { IInjectable } from './Injectable';

interface InjectionTokenOptions<T> {
	scope: ProviderScope;
	factory: () => T;
}

const INJECTION_TOKEN_DEFAULT_OPTIONS = {
	scope: 'root',
	factory: () => undefined,
};

export class InjectionToken<T> {
	readonly metadataName = 'InjectionToken';
	readonly injectable: IInjectable<T>;

	constructor(private readonly _id: string, options: InjectionTokenOptions<T> = INJECTION_TOKEN_DEFAULT_OPTIONS) {
		this.injectable = {
			scope: options.scope,
			factory: options.factory,
			value: undefined,
		};
	}

	toString() {
		const scope = this.injectable.scope;
		const displayPossibleScope = scope && scope !== 'root' ? `@${scope}` : '';

		return `InjectionToken${displayPossibleScope} ${this._id}`;
	}
}
