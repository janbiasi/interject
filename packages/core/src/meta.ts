import 'reflect-metadata';

export const enum TypeDeclaration {
	Component = 'Component',
}

export const metadataKeys = {
	injectable: 'custom:injectable',
	token: 'custom:token',
	provider: 'custom:provider',
	injector: 'custom:injector',
	bootstrap: 'custom:bootstrap',
};

export function getMetadata<T extends any>(metadataKey: any, target: any): T {
	return Reflect.getMetadata(metadataKey, target);
}

export function defineMetadata(metadataKey: any, metadataValue: any, target: any): void {
	Reflect.defineMetadata(metadataKey, metadataValue, target);
}
