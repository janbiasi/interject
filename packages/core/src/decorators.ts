import { IConstructable } from './Constructable';
import { IClassProvider, Provider } from './Provider';
import { InjectionToken } from './InjectionToken';
import { ProviderScope } from './Scope';
import { IInjectableType, IInjectable } from './Injectable';
import { defineMetadata, metadataKeys, getMetadata } from './meta';
import { Injector, CommonInjector } from '.';
import { stringify } from './utils';

export interface IInjectableOptions {
	scope?: ProviderScope;
	requires?: Provider[];
}

const INJECTABLE_DEFAULT_OPTS = {
	scope: 'root',
	requires: [],
};

export function Injectable({ scope, requires }: IInjectableOptions = INJECTABLE_DEFAULT_OPTS) {
	return <T extends { new (...args: any[]): any }>(Factory: T) => {
		const token = new InjectionToken<T>(Factory.name || stringify(Factory));
		const injectable = <IInjectable<T>>{
			scope,
			requires,
			useNew: true,
			factory: Factory as any,
			value: undefined,
		};

		defineMetadata(metadataKeys.token, token, Factory);
		defineMetadata(metadataKeys.injectable, injectable, Factory);
	};
}

export interface IModuleOptions {
	providers?: Array<IConstructable<any> | Provider>;
	parent?: CommonInjector
}

const MODULE_DEFAULT_OPTS = {
	providers: [],
};

/** @internal */
export function internalModuleDecoratorFactory<T extends IConstructable<any>>(
	{ providers, parent }: IModuleOptions,
	Factory: T
): CommonInjector {
	const processedProviders: Provider[] = providers.map(
		(provider: any): Provider => {
			if (typeof provider === 'function') {
				return {
					provide: getMetadata(metadataKeys.token, provider),
					useClass: provider,
				};
			}

			return provider as Provider;
		}
	);

	// create module root injector without parent
	// TODO: let module inherit from other module for correct resolving of inherited deps
	const injector = CommonInjector.create({
		providers: processedProviders,
		parent: parent || CommonInjector.NULL,
		name: Factory.name,
	});

	defineMetadata(metadataKeys.injector, injector, Factory);
	defineMetadata(
		metadataKeys.bootstrap,
		// the tokens in correct order for bootstrap of the module
		processedProviders.map((provider: Provider) => provider.provide),
		Factory
	);

	return injector;
}

export function Module({ providers, parent }: IModuleOptions = MODULE_DEFAULT_OPTS) {
	return <T extends { new (...args: any[]): any }>(Factory: T) => {
		internalModuleDecoratorFactory({
			providers,
			parent
		}, Factory);
	};
}
