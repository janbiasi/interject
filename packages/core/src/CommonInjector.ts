import { IConstructable } from './Constructable';
import { Provider } from './Provider';
import { THROW_IF_NOT_FOUND } from './utils';
import { InjectionToken } from './InjectionToken';
import { InjcetionFlags } from './Flags';
import { IInjectable } from './Injectable';
import { Injector } from './Injector';

// runtime
import { setCurrentInjector, getCurrentInjector } from './runtime';

// core providers
import VersionProvider from './platform/providers/Version';

export interface IInjectorType<T> extends IConstructable<T> {
	injector: IInjector<T>;
}

export interface IInjectorTypeWithProviders<T> {
	pointer: IInjectorType<T>;
	providers?: Provider[];
}

export interface IInjector<T> {
	factory: () => T;
	providers: Provider[];
	imports: (IInjectorType<any> | IInjectorTypeWithProviders<any>)[];
}

export interface ICommonInjectorCreateOptions {
	providers: Provider[];
	parent?: CommonInjector;
	name?: string;
}

export class NullInjector implements CommonInjector {
	get(token: any, notFoundValue: any = THROW_IF_NOT_FOUND): any {
		if (notFoundValue === THROW_IF_NOT_FOUND) {
			throw new Error(`NullInjectorError: No provider for ${token}!`);
		}

		return notFoundValue;
	}
}

export abstract class CommonInjector {
	static NULL: CommonInjector = new NullInjector();
	static THROW_IF_NOT_FOUND = THROW_IF_NOT_FOUND;

	abstract get<T>(token: IConstructable<T> | InjectionToken<T>, notFoundValue?: T, flags?: InjcetionFlags): T;

	static create(options: ICommonInjectorCreateOptions): CommonInjector {
		// add core providers
		options.providers.push(VersionProvider);

		return new Injector(options.providers, options.parent, options.name || null);
	}

	static injectable = <IInjectable<any>>{
		providedIn: 'any',
		factory: () => undefined as any,
		scope: 'root',
		value: undefined,
	};

	static toString() {
		return `CommonInjector`;
	}

	static get currentInjector(): CommonInjector {
		return getCurrentInjector();
	}
}
