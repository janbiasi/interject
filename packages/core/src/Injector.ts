import { CommonInjector } from './CommonInjector';
import { Provider } from './Provider';
import { IInjection } from './InjectionItem';
import { EMPTY, CIRCULAR, stringify } from './utils';
import { InjectionToken } from './InjectionToken';
import { IConstructable } from './Constructable';
import { InjcetionFlags } from './Flags';
import { resolveToken, resolveInjectableProvider } from './resolve';
import { setCurrentInjector } from './runtime';

export class Injector implements CommonInjector {
	readonly parent: CommonInjector;
	readonly source: string | null;

	private _injections: Map<any, IInjection>;

	constructor(providers: Provider[], parent: CommonInjector = CommonInjector.NULL, source: string | null = null) {
		this.parent = parent;
		this.source = source;

		// enable runtime features and closure scoping
		setCurrentInjector(this);

		// storage for all injectables
		const injections = (this._injections = new Map<any, IInjection>());

		// set an injection point to the common injector
		injections.set(CommonInjector, <IInjection>{
			token: CommonInjector,
			factory: CIRCULAR,
			dependencies: EMPTY,
			value: this,
			useNew: false,
		});

		// process all providers
		providers.forEach((provider: Provider) => resolveInjectableProvider(provider, injections));
	}

	set(token: InjectionToken<any> | IConstructable<any>, injection: IInjection): void {
		this._injections.set(token, injection);
	}

	get<T>(token: InjectionToken<any> | IConstructable<any>, notFoundValue?: T, flags?: InjcetionFlags): T {
		const injection = this._injections.get(token);

		// resolve the token and its dependencies
		return resolveToken(token, injection, this._injections, this.parent, notFoundValue, flags);
	}

	toString() {
		const tokens: string[] = [];
		this._injections.forEach((key, token) => tokens.push(stringify(token)));

		return `Injector {Â ${tokens.join(', ')} }`;
	}
}
