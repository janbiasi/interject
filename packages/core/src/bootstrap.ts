import { IConstructable } from './Constructable';
import { getMetadata, metadataKeys } from './meta';
import { CommonInjector } from './CommonInjector';
import { Provide } from './Provider';
import { InjectionToken } from './InjectionToken';

export function bootstrap<T>(ModuleFactory: IConstructable<T>): T {
	const injector: CommonInjector = getMetadata(metadataKeys.injector, ModuleFactory);
	const tokens: (InjectionToken<any> | IConstructable<any>)[] = getMetadata(metadataKeys.bootstrap, ModuleFactory);
	const providers = tokens.map((token: Provide) => injector.get(token));
	const instance = new ModuleFactory(...providers);

	return instance;
}
