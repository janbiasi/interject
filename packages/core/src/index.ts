import { CommonInjector } from './CommonInjector';
import { Injector } from './Injector';
import { InjectionToken } from './InjectionToken';
import { InjcetionFlags, LookupFlags } from './Flags';
import { Provider, IValueProvider, IExistingProvider, IClassProvider, IFactoryProvider } from './Provider';
import { Injectable, Module, internalModuleDecoratorFactory } from './decorators';

export {
	// basic core
	Injector,
	CommonInjector,
	InjectionToken,
	// flags
	InjcetionFlags,
	LookupFlags,
	// providers
	Provider,
	IValueProvider,
	IExistingProvider,
	IClassProvider,
	IFactoryProvider,
	// internal decorator factories
	internalModuleDecoratorFactory,
	// decorators
	Injectable,
	Module,
};
