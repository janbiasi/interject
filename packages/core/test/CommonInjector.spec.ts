import { InjectionToken } from '../src/InjectionToken';
import { Injector } from '../src/Injector';
import { NullInjector, CommonInjector } from '../src/CommonInjector';

const NULL_PROVIDER_TOKEN = new InjectionToken<null>('null-provider');

describe('CommonInjector', () => {
	it('should provide a static NullInjector', () => {
		expect(CommonInjector.NULL).toBeInstanceOf(NullInjector);
	});

	it('should provide a default injectable definition', () => {
		expect(CommonInjector.injectable).toBeDefined();
		expect(CommonInjector.injectable.value).toBe(undefined);
		expect(CommonInjector.injectable.scope).toEqual('root');
		expect(typeof CommonInjector.injectable.factory).toEqual('function');
		expect(CommonInjector.injectable.factory()).toBe(undefined);
	});

	it('should create a single option-ish Injector statically', () => {
		const simpleProviderToken = new InjectionToken<number>('simpleProviderToken');
		const withoutProviders = CommonInjector.create({
			providers: [{ provide: simpleProviderToken, useValue: 20 }],
			name: 'RootInjector',
			parent: CommonInjector.NULL,
		});

		expect(withoutProviders).toMatchSnapshot();
		expect(withoutProviders.toString()).toEqual(
			'Injector {Â CommonInjector, InjectionToken simpleProviderToken, InjectionToken version }'
		);
	});

	describe('NullInjector', () => {
		it('should throw an error in default case', () => {
			const nuller = new NullInjector();

			expect(() => {
				nuller.get(NULL_PROVIDER_TOKEN);
			}).toThrow('NullInjectorError: No provider for InjectionToken null-provider!');
		});

		it('should provide the defalt value given', () => {
			const nuller = new NullInjector();

			expect(nuller.get(NULL_PROVIDER_TOKEN, 10)).toBe(10);
		});

		it('should be used inside every injector as fallback', () => {
			const missingProviderToken = new InjectionToken('missingProviderToken');
			const injector = new Injector([]);

			expect(() => {
				injector.get(missingProviderToken);
			}).toThrow('NullInjectorError: No provider for InjectionToken missingProviderToken!');
		});
	});
});
