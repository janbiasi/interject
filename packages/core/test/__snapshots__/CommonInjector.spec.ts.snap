// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CommonInjector should create a single option-ish Injector statically 1`] = `
Injector {
  "_injections": Map {
    [Function] => Object {
      "dependencies": Array [],
      "factory": [Function],
      "token": [Function],
      "useNew": false,
      "value": [Circular],
    },
    InjectionToken {
      "_id": "simpleProviderToken",
      "injectable": Object {
        "factory": [Function],
        "scope": "root",
        "value": undefined,
      },
      "metadataName": "InjectionToken",
    } => Object {
      "dependencies": Array [],
      "factory": [Function],
      "token": InjectionToken {
        "_id": "simpleProviderToken",
        "injectable": Object {
          "factory": [Function],
          "scope": "root",
          "value": undefined,
        },
        "metadataName": "InjectionToken",
      },
      "useNew": false,
      "value": 20,
    },
    InjectionToken {
      "_id": "version",
      "injectable": Object {
        "factory": [Function],
        "scope": "root",
        "value": undefined,
      },
      "metadataName": "InjectionToken",
    } => Object {
      "dependencies": Array [],
      "factory": [Function],
      "token": InjectionToken {
        "_id": "version",
        "injectable": Object {
          "factory": [Function],
          "scope": "root",
          "value": undefined,
        },
        "metadataName": "InjectionToken",
      },
      "useNew": false,
      "value": "0.1.0",
    },
  },
  "parent": NullInjector {},
  "source": "RootInjector",
}
`;
